controllers/topology/cluster_controller_test.go

```diff
@@ -0,0 +1,377 @@
+
+	// 1) Templates for Machine, Cluster, ControlPlane and Bootstrap.
+	infrastructureMachineTemplate := builder.InfrastructureMachineTemplate(ns.Name, "inframachinetemplate").Build()
+	infrastructureClusterTemplate := builder.InfrastructureClusterTemplate(ns.Name, "infraclustertemplate").
+		// Create spec fake setting to assert that template spec is non-empty for tests.
+		WithSpecFields(map[string]interface{}{"spec.template.spec.fakeSetting": true}).
+		Build()
+	controlPlaneTemplate := builder.ControlPlaneTemplate(ns.Name, "cp1").
+		WithInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		Build()
+	bootstrapTemplate := builder.BootstrapTemplate(ns.Name, "bootstraptemplate").Build()
+
+	// 2) ClusterClass definitions including definitions of MachineDeploymentClasses used inside the ClusterClass.
+	machineDeploymentClass := builder.MachineDeploymentClass(ns.Name, "md1").
+		WithClass(workerClassName).
+		WithInfrastructureTemplate(infrastructureMachineTemplate).
+		WithBootstrapTemplate(bootstrapTemplate).
+		WithLabels(map[string]string{"foo": "bar"}).
+		WithAnnotations(map[string]string{"foo": "bar"}).
+		Build()
+	machineDeploymentClass2 := builder.MachineDeploymentClass(ns.Name, "md1").
+		WithClass(workerClassName2).
+		WithInfrastructureTemplate(infrastructureMachineTemplate).
+		WithBootstrapTemplate(bootstrapTemplate).
+		Build()
+	clusterClass := builder.ClusterClass(ns.Name, clusterClassName).
+		WithInfrastructureClusterTemplate(infrastructureClusterTemplate).
+		WithControlPlaneTemplate(controlPlaneTemplate).
+		WithControlPlaneInfrastructureMachineTemplate(infrastructureMachineTemplate).
+		WithWorkerMachineDeploymentClasses([]clusterv1.MachineDeploymentClass{*machineDeploymentClass, *machineDeploymentClass2}).
+		Build()
+
+	// 3) A Cluster object including a Cluster Topology Object and the MachineDeploymentTopology objects used in the
+	// Cluster Topology.
+	machineDeploymentTopology := builder.MachineDeploymentTopology("mdm").
+		WithClass(workerClassName).
+		WithReplicas(3).
+		Build()
+	machineDeploymentTopology2 := builder.MachineDeploymentTopology("mdm2").
+		WithClass(workerClassName).
+		WithReplicas(1).
+		Build()
+	cluster := builder.Cluster(ns.Name, clusterName).
+		WithTopology(
+			builder.ClusterTopology().
+				WithClass(clusterClass.Name).
+				WithMachineDeployment(machineDeploymentTopology).
+				WithMachineDeployment(machineDeploymentTopology2).
+				WithVersion("1.22.2").
+				WithControlPlaneReplicas(3).
+				Build()).
+		Build()
+
+	// Create a set of initObjects from the objects above to add to the API server when the test environment starts.
+	initObjs := []client.Object{
+		clusterClass,
+		cluster,
+		infrastructureClusterTemplate,
+		infrastructureMachineTemplate,
+		bootstrapTemplate,
+		controlPlaneTemplate,
+	}
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
```

Comment:
nit: I would move it to the first line of this test (as in other cases) to make it directly clear that we're testing with the feature gate enabled and it doesn't interrupt the flow of the code here.